Role arg spec validation and adhoc ansible-role tool (impl of proposal #39)

alikins opened the pull request:
##### SUMMARY
 Start of implementation of https://github.com/ansible/proposals/issues/39 

This adds support for validating role args based on a arg spec provided with the role. The role
arg spec data will be optionally included in a role in a meta/arg_spec.yml. 

The role spec proposal has two main purposes:
- Provide a way for role args to be validated before executing a role
- Expose the role arg information so tooling like Tower can know how to call a role properly and provide user interfaces for ad hoc calling of roles.

A third feature is the addition of a 'ansible-role' tool that allows ad hoc execution of roles in a fashion similar to how 'ansible' ('adhoc') allows for execution of modules.




Assumptions:
 - The role arg spec (the language/data used to define what are valid arguments to provide a role) will be based on ansible module's 'argument_spec' and use the same implementation of the validation and checking. 
- The role arg spec will be provided as a yaml file in a roles meta/ dir (meta/arg_spec.yml). 
- The role arg spec data will be constructed manually to start. In the future, tooling could be provided to partially create this data but that is not covered by this.

Analogs:
When creating a ansible module in python, the typical way is by using ansible_modules.basic.AnsibleModule. The 'argument_spec' to AnsibleModule is a dict that defines the arguments that the module can be called with (the aguments name, type, choices, etc). 

The data in meta/arg_spec.yml is equivalent to AnsibleModule's 'argument_spec' value. In practice, the arg spec data is validated by using the data for the 'argument_spec' arg to AnsibleModule and checking for errors or problems
 
This implementation is partially based on the 'validate_role_spec' feature of ansible network-engine (https://github.com/ansible-network/network-engine). In particular, https://github.com/ansible-network/network-engine/blob/devel/action_plugins/validate_role_spec.py

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, task or feature -->
roles

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.a1.post0 (role_survey 982343ed31) last updated 2018/08/31 13:23:06 (GMT -400)
  config file = None
  configured module search path = ['/home/adrian/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 3.6.6 (default, Jul 19 2018, 16:29:00) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]


```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

A 'cooking' role

roles/cooking/meta/argument_specs.yml
``` yaml
---
# The name of the argspec.
# Multiple arg specs can be defined in argument_specs.yml
# 'main' is used by default if not otherwise specified.
main:
    # A dictionary of the same form as the
    # 'argument_spec' dictionary passed to AnsibleModule.
    # It defines the names of the arguments,
    # and details about the argument.
    argument_spec:

        # The key here is the name of the argument.
        # It's value is a dict.
        # This is equilivent to
        # hostname=dict(type='str', required=True)
        # in a AnsibleModule call
        hostname:
            type: "str"
            required: true

        # port is an optional int arg, defaulting to 8000
        port:
            type: "int"
            default: 8000

        request:
            type: "str"
            choices:
                - "GET"
                - "POST"
                - "PUT"
            required: true

        token:
            type: "str"
            default: null
            no_log: true

        salt:
            type: "str"
            default: "Himalayan"

        pepper:
            type: "str"
            choices:
                - "Tellicherry"
                - "those little packets"

        oil:
            type: "bool"
            required: false
        water:
            type: "bool"
            required: false

    # Other arg spec modifiers, also analog to AnsibleModule args
    mutually_exclusive:
        - ["oil", "water"]

    required_together:
        - ["salt", "pepper"]
```

Example play:
``` yaml

---
- name: "computer help me Im hungry"
  hosts: localhost
  gather_facts: false

  tasks:
    - name: "include_role cooking"
      include_role:
        name: cooking
      vars:
        salt: true
        port: 37.42
        pepper: "green"
        oil: true
        water: true
      register: results_include_role_cooking

    - name: show results_include_role_cooking
      debug:
        var: results_include_role_cooking
```

Example ansible-playbook -vvv output:

``` 
(ansible-py3) [newswoop:F27:cooking]$ ansible-playbook -i hosts test_cooking.yml  -vvv
ansible-playbook 2.8.0.dev0 (role_survey e82f07aeb7) last updated 2018/09/19 14:27:44 (GMT -400)
  config file = None
  configured module search path = ['/home/adrian/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible-playbook
  python version = 3.6.6 (default, Jul 19 2018, 16:29:00) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
No config file found; using defaults
Parsed /home/adrian/src/cooking/hosts inventory source with ini plugin

PLAYBOOK: test_cooking.yml *****************************************************************************************************************************************************************************************
1 plays in test_cooking.yml

PLAY [computer help me Im hungry] **********************************************************************************************************************************************************************************
META: ran handlers

TASK [include_role cooking] ****************************************************************************************************************************************************************************************
task path: /home/adrian/src/cooking/test_cooking.yml:7

TASK [cooking : Validating arguments against arg spec 'main'] ******************************************************************************************************************************************************
task path: /home/adrian/src/cooking/test_cooking.yml:2
fatal: [localhost]: FAILED! =>
```
``` json
 {
    "argument_errors": [
        "parameters are mutually exclusive: oil, water",
        "missing required arguments: hostname, request",
        "argument port is of type <class 'float'> and we were unable to convert to int: <class 'float'> cannot be converted to an int",
        "value of pepper must be one of: Tellicherry, those little packets, got: green"
    ],
    "argument_spec_data": {
        "argument_spec": {
            "hostname": {
                "required": true,
                "type": "str"
            },
            "oil": {
                "required": false,
                "type": "bool"
            },
            "pepper": {
                "choices": [
                    "Tellicherry",
                    "those little packets"
                ],
                "type": "str"
            },
            "port": {
                "default": 8000,
                "type": "int"
            },
            "request": {
                "choices": [
                    "GET",
                    "POST",
                    "PUT"
                ],
                "required": true,
                "type": "str"
            },
            "salt": {
                "default": "Himalayan",
                "type": "str"
            },
            "token": {
                "default": null,
                "no_log": true,
                "type": "str"
            },
            "water": {
                "required": false,
                "type": "bool"
            }
        },
        "mutually_exclusive": [
            [
                "oil",
                "water"
            ]
        ],
        "required_together": [
            [
                "salt",
                "pepper"
            ]
        ]
    },
    "changed": false,
    "msg": "Validation of arguments failed:\nparameters are mutually exclusive: oil, water\nmissing required arguments: hostname, request\nargument port is of type <class 'float'> and we were unable to convert to int: <class 'float'> cannot be converted to an int\nvalue of pepper must be one of: Tellicherry, those little packets, got: green",
    "validate_args_context": {
        "argument_spec_name": "main",
        "name": "cooking",
        "path": "/home/adrian/src/cooking/roles/cooking",
        "type": "role"
    }
}
	to retry, use: --limit @/home/adrian/src/cooking/test_cooking.retry

PLAY RECAP *********************************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

```

Note in particular the 'argument_errors':
``` json
[
        "parameters are mutually exclusive: oil, water",
        "missing required arguments: hostname, request",
        "argument port is of type <class 'float'> and we were unable to convert to int: <class 'float'> cannot be converted to an int",
        "value of pepper must be one of: Tellicherry, those little packets, got: green"
 ]
```

And the 'validate_args_context' with info about what was being called with invalid args:

``` json
 {
        "argument_spec_name": "main",
        "name": "cooking",
        "path": "/home/adrian/src/cooking/roles/cooking",
        "type": "role"
}
```

alikins committed:
Add validate_arg_spec action/module

Add integration tests for validate_arg_spec

Update to use AnsibleModule arg spec handling

Now returns a list of all argument errors not
just the first one (in the results['argument_errors']
value).

Add a role with a meta/arg_specs.yml

Add a kwarg for 'validate_args_context' to pass in a dict
with info about what context (ie, if role loader is doing it
then include some info about the role like the role name and path).

Also pass in full 'argument_spec_data' to the action, so
that we can include that in the result info as well.

Add unit test for validate_arg_spec action

alikins committed:
Let the arg spec validate task 'name' be set from data

This servces two purposes:
 1) Allow roles to add additional info about the arg spec
    that will show up in the default output.
    ie, 'Validating arg spec all_the_clouds has all the required cloud
    args'

 2) let test cases change the name of the injected implicit
    validate_arg_spec task so they can include 'EXPECTED FAILURE'
    in the name so ansible-test's junit_callback doesnt set the
    task as failed when it is expected to fail

Default to style like:

    TASK [test1 : Validating arguments against arg spec 'main']

If the argument_spec includes a 'description' field, then it gets
added to the task name like:

    TASK [test1 : Validating arguments against arg spec \
      'main' - The argument spec for the 'test1' role]

Add argument_spec_name to the validate_args_context data.

alikins committed:
Add argument_spec info to Role.[de]serialize

Add unit tests for Roles.[de]serialize

And add a more complete role example to
also cover things like compile(),
get_vars(), get_default_vars(), get_handler_blocks()

alikins committed:
validate_arg_specs intg test updates

Add runme.sh for intg tests
Sort expected argument_errors and results for assert
Finer grained asserts
Clean up intg test, check for set diff in actual/expected

Fix intg tests failures between py2/py3 arg valid messages

Replace any py2/py3 specific type/class reprs (like <type 'unicode'>)
with generic names so we match when we do set compares.

Update test_validate_role_spec (network-engine compat tests)

Add some pre/post tasks in intg tests

Use 'other' alt name for argspec info

Test we are using facts for task_vars correctly

Skip the validate_arg_spec test if jinja < 2.7 (ie, centos6)

The validate_arg_spec test_include_role_fails intg test uses
the jinja 'map' filter introduced in jinja 2.7.
So skip the test if we have jinja 2.6 or earlier.

ignore_errors on the include_role block

rm unused test roles

alikins committed:
Start loading meta/argument_specs.yml on role load

Insert arg_spec validation task in task_data if arg_spec

If a role has a meta/argument_specs.yml with an applicable
argument_spec item in it, create a task datastructure
that represents an invocation of the validate_arg_specs action
using that argument_spec data.

Do this even if there are no other tasks in the role.

Hardcode role argument_spec_name to 'main'

delog role/__init__.py

alikins commented:
(moved from description)

TODO:
- [x] add a validate_arg_spec action plugin
- [x] look for and load a roles meta/argument_spec.yml
- [x] modify role loading to inject a validate_arg_spec task before role use
- [x] REQUIRED - ansible-role cli - update to only use Role level arg spec checking
- [x] REQUIRED - cleanup/remove debug logging
- [x] REQUIRED - delete playbook.role.survey stuff
- [x] REQUIRED - intg tests - ansible-role
- [ ] REQUIRED - squash/rebase/cleanup git history
- [x] NICE - unit tests - playbook.role changes
- [x] NICE - unit tests - port network-engine test cases
- [x] NICE - unit tests - ansible-role?
- [x] NICE - unit tests - validate_arg_spec action
- [x] NICE - intg tests - port network-engine validate_role_spec intg tests
- [x] NICE - intg tests - validate_arg_spec
- [x] NICE - intg tests - validate_arg_spec - cover the different arg spec arg types
- [x] NICE - intg tests - validate_arg_spec - cover the various 'required_if', 'mutually_exclusive' etc
- [x] NICE - intg tests - include_role (including from_vars/from_defaults?)
- [x] NICE - intg tests - import_role 
- [x] NICE - intg tests - play level roles  (plain names, json like dicts, maybe with kw style args)
- [ ] MAYBE - play/task keyword for specifying which named argument_spec to use (if different than 'default')
- [ ] MAYBE - play/task keyword to enable/disable arg spec checking even when a argument_specs.yml exists 
- [ ] MAYBE - implement a playbook.role.arg_spec.ArgSpec
- [x] MAYBE - Return more structured info from validate_arg_spec for error reporting
- [x] MAYBE - Add some metadata to validate_arg_spec calls to be used in errors
- [ ] MAYBE - validate_arg_spec error details - path to the file where the argument_spec came from (if it came from a file)
- [ ] MAYBE - validate_arg_spec error details - add a 'source' param?
- [x] MAYBE - validate_arg_spec error details - pass in role version? role name?
- [ ] MAYBE - validate_arg_spec error details - pass in argument_specs.yml shasum?
- [x] MAYBE - validate_arg_spec error details - pass in some 'context' info (ie, a note that the action is being called to check role specs vs task specs since it could in theory check task specs as well)
- [ ] MAYBE - ui/uxd - tweak default callback to show arg spec failures nicely
- [ ] MAYBE - ui/uxd - tweak module_utils.basic.ModuleUtils self._check_* methods to pass more info to fail_json so we can use it in error messages
- [ ] MAYBE - update names/terminology ('arg spec') etc if needed
- [x] MAYBE - docs - validate_arg_spec action
- [ ] MAYBE - docs - argument_specs.yml format and use
- [ ] MAYBE - docs - ansible-role cli

sivel commented on lib/ansible/modules/utilities/logic/validate_arg_spec.py:
```diff
@@ -0,0 +1,74 @@
+
+DOCUMENTATION = '''
+---
+module: validate_arg_spec
```
We refer to the `argument_spec` in shorthand as `arg_spec` a lot, but I feel to remain consistent, this should use the full `argument` naming.

sivel replied:
The same goes with any use of the shorthand `arg` or `args`.  We should spell this out.

bcoca commented:
To bikeshed on naming and in order of personal preference:
 - validate_variables: strickly speaking this is what it is doing, as roles don't take arguments but inherit variables
 - validate_arguments: kind of wrong, but most people will understand what it means
 - validate_spec: turns it around to validate we conform to the specification
 - validate_argument_spec: a bit long, also imprecise, but good enough

bcoca commented on lib/ansible/modules/utilities/logic/validate_arg_spec.py:
```diff
@@ -0,0 +1,74 @@
+    description:
+        - A dictionary like AnsibleModule argument_spec
+    required: true
+  provided_arguments:
```
we should not need to provide arguments, the 'current variables' are what we want to check against

see example on use with `role_params` below for specifics

bcoca commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -222,7 +223,45 @@ def _load_role_data(self, role_include, parent_role=None):
+        # TODO: need a playbook.base.Base derived object here?
+        # TODO: do we want a Role (or Task or Base) object to have a arg_spec attribute?
+        #       Seems like it would be handy for introspection and error reporting and copy()
+        self._argument_specs = argument_specs
```
makes sense to hang it on the role, since that is what we are checking in this case

bcoca commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -222,7 +223,45 @@ def _load_role_data(self, role_include, parent_role=None):
+
         task_data = self._load_role_yaml('tasks', main=self._from_files.get('tasks'))
+
+        argument_spec_name = 'main'
```
as discuss we expect this to match 'tasks_from', which defaults to main when not provided (its only optionally provided on include_role)

bcoca commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -281,6 +321,31 @@ def _load_dependencies(self):
+                         # 'vars': {'argument_spec': []},
+                         # 'async_val': async_val,
+                         # 'poll': poll},
+                         }
```
yes, we should inherit all properties ... not sure if we'll eventually want a 'ignore or skip spec validation' option for roles

alikins replied:
I'll remove the TODO, since at the moment it inherits the properties.

bcoca commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -281,6 +321,31 @@ def _load_dependencies(self):
+
+        arg_spec_task = {'action': {'module': 'validate_arg_spec',
+                                    'argument_spec': argument_spec,
+                                    'provided_arguments': role_params,
```
role_params are not the only things valid, any variable provided is, just because a required variable is not provided as 'role_param' does not mean the role wont work. i.e
```
roles:
   - name: myrole
     thisisroleparam: othervalue
     vars:
        notroleparambutvalid: value
```
this should work with BOTH variables being required

alikins replied:
That works currently, added https://github.com/ansible/ansible/pull/44983/files#diff-c8709e49ceb22733ac4acc130e841708 as a test and example.

bcoca commented on lib/ansible/playbook/role_include.py:
```diff
@@ -131,7 +131,10 @@ def load(data, block=None, role=None, task_include=None, variable_manager=None,
-            raise AnsibleParserError('Invalid options for %s: %s' % (ir.action, ','.join(list(bad_opts))), obj=data)
+            raise AnsibleParserError('Invalid options for %s: %s valid options are: (%s)' % (ir.action,
+                                                                                             ','.join(list(bad_opts)),
+                                                                                             ','.join(list(IncludeRole.VALID_ARGS))),
```
i would not display 'valid_args' here as they are not 'valid' in all cases of role invocation, this code gets used for both import/include_role and `roles:`

alikins replied:
Hmm, indeed. Removing for now. Maybe knowing/showing valid options for both cases is something to address in a different pr...

bcoca commented on lib/ansible/plugins/action/validate_arg_spec.py:
```diff
@@ -0,0 +1,153 @@
+        argument_spec = argument_spec_data.pop('argument_spec', [])
+
+        # the values that were passed in and will be checked against argument_spec
+        provided_arguments = self._task.args.get('provided_arguments', {})
```
this should just use task_vars as commented above, no need to pass a specific argument

alikins replied:
provided_args allows a task to pass in an explicit data structure to validate without it needing to exist somewhere in ansible vars.

bcoca replied:
'normal use case' is validating 'current vars', which user cannot really pass in, they can only acces subsets of 'task_vars' ... may be ok as option (what use case?) but it should at least default to task_vars .. which is what we need to validate in the case from the proposal

abadger commented:
@nitzmahone and I talked about this after last week's meeting.  We came to the conclusion that the action plugin is useful to users but should not be involved in the implicit ansible validation of role specs.  We felt the architecture should look something like:

* We should have a library call for arg_spec for multiple reasons which will become apparent in the following bullets.  Currently the code is inside of basic.py::AnsibleModule() but that's kinda icky because of being inside of an unrelated module.
  * nitzmahone thought that we should work on finishing sivel's extraction of arg_spec as the way to make that sanely available as a library
  * I agree that the extraction is the best way forward but if we're in a time crunch for this, I could see taking alikins' work to fake out the AnsibleModule stuff and turning that into our library function temporarily.  (Maybe as a private method so that nothing outside of our tree tries to use it?)
* nitzmahone didn't think the role validating code should be overridable... I'm leaning towards that as well after listening to his arguments.  In addition, I think calling the action plugin from ansible role interpretation code is probably overcomplicating things. So we think this should be changed to calling the library function directly.
* Separately, having an action plugin *is* useful for people to be able to validate their own data structures.  We talked a bit about whether an action plugin is the proper way to do this.... I feel that it makes more sense to compose this from a single jinja call plus the assert module while nitzmahone feels that pushing all the elements into a single action plugin, as done here, makes more sense.  he swayed me to his point of view, however, by pointing out that our assert module is very weak on showing why an assert failed.  We would want something similar to pytest which expands the vars for good error messages.  Assert can't do this without a good deal more code (We'd probably have to implement code similar to pytest's assert rewriting logic to do this.  It wouldn't need to search for asserts but it would need to parse the conditions similarly to how pytest does it: http://pybites.blogspot.com/2011/07/behind-scenes-of-pytests-new-assertion.html ) .  Since the action plugin is not part of role_spec validation in this scheme, it probably should go into a separate PR which uses the library functions implemented here.
* If other people want to write a jinja2 test which can be used to validate their data then they can make use of the same library functions implemented above. If we take the private method route above, we'd need to port these to use the permanent solution once it becomes available.
* Somewhat of a tangent, fully qualified namespaces for modules would let us have role code as an action plugin that is not overridable.  However, that seems a round-about and overcomplicated way to achieve the functionality when we have a library function that can do it directly.

nitzmahone also floated the alternative idea that  the validation could be explicit rather than implicit (The role spec and the data have to be available to the playbook.  Then the playbook author explicitly calls the validation action plugin on the data in one task, then passes it to the role in the next task.  I'm not as in favour of that approach, but maybe someone else thinks it makes more sense?

alikins committed:
Try to use arg spec name matching tasks filename for from_tasks

If a role is loaded using from_tasks, the validate arg spec task
will use that file name to lookup a arg spec in the dict in
meta/argument_specs.yml.

For example, a include_role with a "task_from: other.yml", the
args to the role will be validated with the arg spec data
under the 'other.yml' key in argument_specs.yml. If there is
no 'other.yml' key in the dict, then 'main' will be tried.
If there is no 'main' in the dict, arg spec validation will
be skipped.

alikins committed:
Add test of arg spec validating against vars for roles

alikins commented:
@abadger 

> "We should have a library call for arg_spec for multiple reasons which will become apparent in the following bullets. "

> nitzmahone thought that we should work on finishing sivel's extraction of arg_spec as the way to make that sanely available as a library

Agreed in principle (see https://github.com/ansible/ansible/compare/devel...alikins:arg_spec for my take on it). 
But... we don't have that currently.

alikins commented:
> We would want something similar to pytest which expands the vars for good error messages. Assert can't do this without a good deal more code (We'd probably have to implement code similar to pytest's assert rewriting logic to do this. It wouldn't need to search for asserts but it would need to parse the conditions similarly to how pytest does it: 

Something like https://gist.github.com/alikins/683b1bc8a3105d38d5f38fdb65a14dcd ?

privateip commented:
@abadger @nitzmahone I competely understand the rational for not wanting to do this as an action plugin.  Please keep in mind though that roles now potentially have multiple entry points and that depending on the entry point, the validation rules could be different.

ant31 commented:
I would suggest using OpenAPiv3 as the spec? there is already many tooling around this spec (to generate documentation, validation...). It's used by many other projects to define API and then already known by many users.

lwm commented:
Nice work on this @alikins. At this point, I'm not sure what the status is. This PR is 1,400+ LOC and seems to be tackling 3 things (the spec, the tower integration and the ansible-role tool) so it is very hard to know how to get involved to push it forward. It appears to have stalled so far? I'm going to take this to the core meeting tomorrow and see what happens.

Shrews commented:
Closing this. Role arg spec implemented in PR #73152.
